openglyph/src/utility/steam_paths.cpp

```diff
@@ -0,0 +1,119 @@
+#include "vdf_fileparser.hpp"
+
+#include <khepri/io/exceptions.hpp>
+#include <khepri/utility/platform.hpp>
+
+#include <openglyph/utility/steam_paths.hpp>
+
+#include <iostream>
+namespace openglyph::utility {
+
+/// @brief Returns the usual Steam root installation path for the current platform.
+/// @return The root Steam installation directory as a std::filesystem::path.
+std::filesystem::path get_steam_root_path()
+{
+    switch (khepri::utility::get_current_platform()) {
+    case khepri::utility::Platform::windows:
+#if defined(_WIN32) || defined(_WIN64)
+    {
+        std::filesystem::path steampath =
+            khepri::utility::get_registry_key(HKEY_LOCAL_MACHINE,
+                                              "SOFTWARE\\WOW6432Node\\Valve\\Steam", "InstallPath")
+                .value_or("C:\\Program Files (x86)\\steam");
+        if (!std::filesystem::exists(steampath)) {
+            throw khepri::io::FileNotFoundError();
+        }
+
+        if (!std::filesystem::exists(steampath / "steam.exe")) {
+            throw khepri::io::FileNotFoundError();
+        }
+        return steampath;
+    }
+#else
+        return "C:\\Program Files (x86)\\steam";
+#endif
+    case khepri::utility::Platform::linux: {
+        std::filesystem::path steampath = "~/.steam";
+
+        if (!std::filesystem::exists(steampath)) {
+            throw khepri::io::FileNotFoundError();
+        }
+
+        if (!std::filesystem::exists(steampath / "steam")) {
+            throw khepri::io::FileNotFoundError();
+        }
+        return steampath;
+    }
+    }
+    throw khepri::Error("platform not supported");
+}
+
+/// @brief Returns the path to the Steam 'libraryfolders.vdf' file, which contains additional
+/// Steam library locations.
+/// @return The full path to the 'libraryfolders.vdf' file as a std::filesystem::path.
+std::filesystem::path get_steam_library_folders_path()
+{
+    return get_steam_root_path() / "steamapps" / "libraryfolders.vdf";
+}
+
+/// @brief Returns the folder locations for all the steam library folders
+/// @return The vector containing all the library folder paths
+std::vector<std::filesystem::path> get_steam_library_folders()
+{
+    const std::filesystem::path libraryfolders_path = get_steam_library_folders_path();
+
+    if (!std::filesystem::exists(libraryfolders_path)) {
+        throw khepri::io::FileNotFoundError();
+    }
+
+    std::ifstream file(libraryfolders_path);
+    if (!file.is_open()) {
+        throw khepri::io::Error("Could not open file");
+    }
+
+    const auto root = tyti::vdf::read(file);
```

Comment:
This throws `std::runtime_error` if parsing failed. We should catch that and re-throw as an `InvalidFormatError`.